### PR TITLE
Feature/google calendar admin

### PR DESCRIPTION
**Pull request by code-ph0y**
Move Google Calendar Integration from System Settings to Administration section.
 
## Description

The current google calendar authentication is located within the System Settings. This pull request will take the google calendar authentication from the System Settings section and put it within its own Administration section (Google Calendar).

## Motivation and Context
This change is to allow the feature to be located more easily for the user

## How To Test This

1. Go to: Admin section.
2. Locate "Google Calendar" Section
3. Click "Authentication" Link
4. Select Google Calendar Authentication JSON File
5. Click Save
6. Re-enter "Authentication" link Configured should be Green

7. Google Calendar integration should work as expected

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [ ] My change requires a change to the documentation.
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

File: modules/Configurator/controller.php [hunk 1]
```diff
@@ -45,6 +45,49 @@
...
+        require_once('modules/Administration/QuickRepairAndRebuild.php');
+        global $current_user;
+        if(!is_admin($current_user)){
+            sugar_die($GLOBALS['app_strings']['ERR_NOT_ADMIN']);
+        }
+        $configurator = new Configurator();
+
+        if ($configurator->saveConfig() === false)
+        {
+            echo 'saveConfig returned false';
```
Comment (gymad) [inline modules/Configurator/controller.php]: Use log or translated user message instead native echo.

File: modules/Configurator/controller.php [hunk 2]
```diff
@@ -45,6 +45,49 @@
 require_once('include/MVC/Controller/SugarController.php');
 class ConfiguratorController extends SugarController
 {
+
+    /**
+     * Go to the authentication view
+     */
+    function action_gcauthentication(){
```
Comment (gymad) [inline modules/Configurator/controller.php]: Use the functions visibility (`public`), use PSR-2 coding standard on php files.

File: modules/Administration/language/en_us.lang.php [hunk 3]
```diff
@@ -801,6 +801,12 @@
     'LBL_JJWG_MAPS_ADMIN_ADDRESS_CACHE_DESC' => 'Provides access to Address Cache information. This is only cache.',
 
 
+    // Google auth
```
Comment (horus68) [inline modules/Administration/language/en_us.lang.php]: 2 Google auth comments (see also line 989)

**Comment by samus-aran** [comment]
Comments given.

File: modules/Administration/GoogleCalendarAuth.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+* reasonably feasible for technical reasons, the Appropriate Legal Notices must
+* display the words "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+*/
+
+if (!defined('sugarEntry') || !sugarEntry) {
+   die('Not A Valid Entry Point');
+}
+
+
+include 'modules/Configurator/Configurator.php';
```
Comment (gymad) [inline modules/Administration/GoogleCalendarAuth.php]: Use absolute path with `__DIR__` constant before filenames.

File: modules/Administration/GoogleCalendarAuthHandler.php [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        $this->modStrings   = $mod_strings;
+        $this->ss           = $sugar_smarty;
+        $this->tplPath      = $tpl_path;
+        $this->js           = $js;
+        $this->configurator = $config;
+
+        if (isset($this->request['do']) && $this->request['do'] == 'save') {
+            $this->configurator->config['google_auth_json'] = !empty($this->request['google_auth_json']);
+            $this->configurator->saveConfig();
+            header('Location: index.php?module=Administration&action=index');
```
Comment (gymad) [inline modules/Administration/GoogleCalendarAuthHandler.php]: There is a function called `SugarApplication::redirect()`, you can use it. But please take out the redirection to a separated non-private function. It makes possible to write mock object in tests later..

File: modules/Administration/GoogleCalendarAuthHandler.php [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+     *
+     * @return void
+     */
+    protected function handleDisplay()
+    {
+        $errors = array();
+
+        // Check current user is admin
+        if (!is_admin($this->currentUser)) {
+            sugar_die("Unauthorized access to administration.");
```
Comment (gymad) [inline modules/Administration/GoogleCalendarAuthHandler.php]: can you please separate out to a non-private method the die(), sugar_die() and exit calls.

File: modules/Administration/GoogleCalendarAuthHandler.php [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+     *
+     * @param  User         $current_user
+     * @param               $request
+     * @param               $mod_strings
+     * @param  Configurator $config
+     * @param  Sugar_Smarty $sugar_smarty
+     * @param  javascript   $js
+     * @return void
+     */
+    public function __construct(string $tpl_path, User $current_user, $request, $mod_strings, Configurator $config, Sugar_Smarty $sugar_smarty, javascript $js)
```
Comment (gymad) [inline modules/Administration/GoogleCalendarAuthHandler.php]: Please do not use primitive types in typehinting by backward compatibility reason. (remove the `string` here)

File: modules/Administration/GoogleCalendarAuthHandler.php [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+     */
+    protected $js = null;
+
+
+
+    /**
+     * Setup object
+     *
+     * @param  User         $current_user
+     * @param               $request
```
Comment (gymad) [inline modules/Administration/GoogleCalendarAuthHandler.php]: missiing types in phpdoc

File: modules/Administration/GoogleCalendarAuth.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+include 'modules/Administration/GoogleCalendarAuthHandler.php';
+
+global $current_user;
+global $mod_strings;
+global $app_strings;
+
+$tplPath = 'modules/Administration/GoogleCalendarAuth.tpl';
+$request = $_REQUEST;
+
+new GoogleCalendarAuthHandler(
```
Comment (gymad) [inline modules/Administration/GoogleCalendarAuth.php]: `GoogleCalendarAuthHandler` as a new class needs unit tests. Can you add test to this class please.

File: modules/Administration/AdminHandlerInterface.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
+<?php
```
Comment (gymad) [inline modules/Administration/AdminHandlerInterface.php]: This file could be deleted.

File: modules/Administration/BaseHandler.php [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     *
+     * @var array
+     */
+    protected $modStrings = null;
+
+    /**
+     * Setup Object
+     */
+    public function __construct(Sugar_Smarty $sugar_smarty, User $current_user, $request, $mod_strings)
```
Comment (gymad) [inline modules/Administration/BaseHandler.php]: missing phpdoc

File: modules/Administration/GoogleCalendarSettingsHandler.php [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+     * @param array        $mod_strings
+     * @param Configurator $config
+     * @param Sugar_Smarty $sugar_smarty
+     * @param javascript   $js
+     */
+    public function __construct($tpl_path, User $current_user, $request, $mod_strings, Configurator $config, Sugar_Smarty $sugar_smarty, javascript $js)
+    {
+        // Get parent
+        parent::__construct($sugar_smarty, $current_user, $request, $mod_strings);
+
```
Comment (gymad) [inline modules/Administration/GoogleCalendarSettingsHandler.php]: Call the parent methods here (as we discussed)

File: modules/Administration/GoogleCalendarSettingsHandler.php [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    /**
+     * Deal with do actions
+     *
+     * @return void
+     */
+    protected function doActions()
+    {
+        if (isset($this->request['do']) && $this->request['do'] == 'save') {
+            echo 'this working';
```
Comment (gymad) [inline modules/Administration/GoogleCalendarSettingsHandler.php]: echo?

File: tests/unit/codecept/modules/Administration/BaseHandlerTest.php [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+class BaseHandlerTest extends SuiteCRM\StateCheckerUnitAbstract
+{
+    public function setUp() {
+        parent::setUp();
+    }
+
+    public function testFirst() {
+        $this->assertEquals(true, true);
```
Comment (gymad) [inline tests/unit/codecept/modules/Administration/BaseHandlerTest.php]: test is missing

File: tests/unit/codecept/modules/Administration/GoogleCalendarSettingsHandlerTest.php [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+class GoogleCalendarSettingsHandlerTest extends SuiteCRM\StateCheckerUnitAbstract
+{
+    public function setUp() {
+        parent::setUp();
+    }
+
+    public function testFirst() {
+        $this->assertEquals(true, true);
```
Comment (gymad) [inline tests/unit/codecept/modules/Administration/GoogleCalendarSettingsHandlerTest.php]: test is missing

File: modules/Administration/BaseHandler.php [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    /**
+     * Get Languages
+     *
+     * @return void
+     */
+    protected function getLanguage()
+    {
+        $this->ss->assign('LANGUAGES', get_languages());
```
Comment (gymad) [inline modules/Administration/BaseHandler.php]: For testing:

and call: $this->ss->assign('LANGUAGES', $this->getLanguage());

protected function getLanguage() {
  return get_language();
}

(anf same with the get_set_focus_js() function)

**Comment by samus-aran** [comment]
Assessed :+1:

File: modules/Administration/language/en_us.lang.php [hunk 17]
```diff
@@ -785,21 +785,28 @@
...
 
-    'LBL_JJWG_MAPS_ADMIN_HEADER' => 'Google Maps',
-    'LBL_JJWG_MAPS_ADMIN_DESC' => 'Manage your geocoding, testing geocoding, view geocoding result totals and configure advanced settings.',
-    'LBL_JJWG_MAPS_ADMIN_CONFIG_TITLE' => 'Google Maps Settings',
-    'LBL_JJWG_MAPS_ADMIN_CONFIG_DESC' => 'Configuration settings to adjust your Google Maps',
-    'LBL_JJWG_MAPS_ADMIN_GEOCODED_COUNTS_TITLE' => 'Geocoded Counts',
-    'LBL_JJWG_MAPS_ADMIN_GEOCODED_COUNTS_DESC' => 'Shows the number of module objects geocoded, grouped by geocoding response',
-    'LBL_JJWG_MAPS_ADMIN_DONATE_TITLE' => 'Donate to this Project',
-    'LBL_JJWG_MAPS_ADMIN_DONATE_DESC' => 'Please consider donating to this project!',
+    'LBL_GOOGLE_SUITE_ADMIN_HEADER'               => 'Google Suite',
```
Comment (horus68) [inline modules/Administration/language/en_us.lang.php]: Please remove spaces before `' => '`

Comment (cameronblaikie) [reply modules/Administration/language/en_us.lang.php]: Hi @horus68 I have added a pull request to resolve this, please check to make sure this is working.

**Pull request merged**